src/mscorlib/src/System/Text/EncodingForwarder.cs

```diff
@@ -192,6 +192,358 @@ public unsafe static int GetBytes(Encoding encoding, char* chars, int charCount,
+                    char* input = pInput + charIndex;
+                    byte* output = pOutput + byteIndex;
+                    lengthEncoded = GetBytesAsciiFastPath(input, output, Math.Min(charCount, byteCount));
+                    if (lengthEncoded < byteCount) {
+                        // Not all ASCII, use encoding's GetBytes for remaining conversion
+                        lengthEncoded += encoding.GetBytesFallback(input + lengthEncoded, charCount - lengthEncoded, output + lengthEncoded, byteCount - lengthEncoded, null);
+                    }
+                }
+            } else {
+                // Nothing to encode
+                lengthEncoded = 0;
+            }
+
+            return lengthEncoded;
+        }
+
+        private unsafe static byte[] GetBytesAsciiFastPath(Encoding encoding, char* input, int charCount)
+        {
+            // Fast path for pure ASCII data for ASCII and UTF8 encoding
+            int asciiLength;
+            int remaining = 0;
+            // Assume string is all ASCII and size array for that
+            byte[] bytes = new byte[charCount];
+
+            fixed (byte* output = &bytes[0]) {
+                asciiLength = GetBytesAsciiFastPath(input, output, charCount);
+                if (asciiLength < charCount) {
+                    // Not all ASCII, get the byte count for the remaining encoded conversion
+                    remaining = encoding.GetByteCount(input + asciiLength, charCount - asciiLength, null);
+                }
+            }
+
+            if (remaining > 0) {
+                // Not all ASCII, fallback to slower path for remaining encoding
+                var encoded = ResizeGetRemainingBytes(encoding, input, charCount, ref bytes, asciiLength, remaining);
+                Debug.Assert(encoded == remaining);
+            }
+
+            return bytes;
+        }
+
+        internal unsafe static int ResizeGetRemainingBytes(Encoding encoding, char* chars, int charCount, ref byte[] bytes, int alreadyEncoded, int remaining)
+        {
+            // Resize the array to the correct size
+            Array.Resize(ref bytes, alreadyEncoded + remaining);
+
+            int encoded;
+            fixed (byte* output = &bytes[0]) {
+                // Use encoding's GetBytes for remaining conversion
+                encoded = encoding.GetBytesFallback(chars + alreadyEncoded, charCount - alreadyEncoded, output + alreadyEncoded, remaining, null);
+            }
+
+            return encoded;
+        }
+
+        internal unsafe static int GetBytesAsciiFastPath(char* input, byte* output, int byteCount)
+        {
+            const int Shift16Shift24 = (1 << 16) | (1 << 24);
+            const int Shift8Identity = (1 <<  8) | (1);
+
+            // Encode as bytes upto the first non-ASCII byte and return count encoded
+            int i = 0;
+#if BIT64
+            if (byteCount < 4) goto trailing;
+
+            int unaligned = (unchecked(-(int)input) >> 1) & 0x3;
```

Comment:
> int unaligned = (unchecked(-(int)input) >> 1) & 0x3; [](start = 11, length = 53)

would be clearer if we write this as 
 int unaligned = (((ulong)input) & 0x7) >> 1;